atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ClusterMetadataUtils.java

```diff
@@ -132,37 +137,46 @@ private static InetSocketAddress lookUpAddress(Map<String, InetSocketAddress> no
         });
         return KeyedStream.stream(tokenMapping.asMap())
                 .map(Collection::stream)
-                .map(stream -> stream.collect(Collectors.toList()))
+                .map(stream -> (RangeSet<LightweightOppToken>) stream.collect(toImmutableRangeSet()))
                 .collectToMap();
     }
 
-    // VisibleForTesting
-    public static Set<TokenRange> getMinimalSetOfRangesForTokens(
-            Metadata metadata, Set<Token> partitionKeyTokens, SortedMap<Token, TokenRange> tokenRangesByEnd) {
-        Map<Token, TokenRange> tokenRangesByStartToken = new HashMap<>();
-        for (Token token : partitionKeyTokens) {
-            TokenRange minimalTokenRange = findTokenRange(metadata, token, tokenRangesByEnd);
+    @VisibleForTesting
+    static RangeSet<LightweightOppToken> getMinimalSetOfRangesForTokens(
+            Set<LightweightOppToken> partitionKeyTokens,
+            SortedMap<LightweightOppToken, Range<LightweightOppToken>> tokenRangesByEnd) {
+        Map<LightweightOppToken, Range<LightweightOppToken>> tokenRangesByStartToken = new HashMap<>();
+        for (LightweightOppToken token : partitionKeyTokens) {
+            Range<LightweightOppToken> minimalTokenRange = findTokenRange(token, tokenRangesByEnd);
             tokenRangesByStartToken.merge(
-                    minimalTokenRange.getStart(), minimalTokenRange, ClusterMetadataUtils::findLatestEndingRange);
+                    LightweightOppToken.getLowerExclusive(minimalTokenRange),
+                    minimalTokenRange,
+                    ClusterMetadataUtils::findLatestEndingRange);
         }
-        return ImmutableSet.copyOf(tokenRangesByStartToken.values());
+        return tokenRangesByStartToken.values().stream().collect(toImmutableRangeSet());
     }
 
-    private static TokenRange findTokenRange(
-            Metadata metadata, Token token, SortedMap<Token, TokenRange> tokenRangesByEnd) {
+    private static Range<LightweightOppToken> findTokenRange(
+            LightweightOppToken token, SortedMap<LightweightOppToken, Range<LightweightOppToken>> tokenRangesByEnd) {
         if (tokenRangesByEnd.containsKey(token)) {
             return tokenRangesByEnd.get(token);
         } else if (!tokenRangesByEnd.headMap(token).isEmpty()) {
-            return metadata.newTokenRange(tokenRangesByEnd.headMap(token).lastKey(), token);
+            // handle wraparound
+            LightweightOppToken lowerBound = tokenRangesByEnd.headMap(token).lastKey();
+            if (lowerBound.toString().isEmpty()) {
```

Comment:
We probably want a more direct isEmpty() check here - this encodes the bytes into a string...